### PR TITLE
Apply DomainTranslator to STARTS_WITH function

### PR DESCRIPTION
**Pull request by takezoe**

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/sql/planner/DomainTranslator.java [hunk 1]
```diff
@@ -928,7 +931,55 @@ protected ExtractionResult visitLikePredicate(LikePredicate node, Boolean comple
...
+            return visitExpression(node, complement);
+        }
+
+        private Optional<ExtractionResult> tryVisitStartsWithFunction(FunctionCall node, Boolean complement)
+        {
+            List<Expression> args = node.getArguments();
+            if (args.size() != 2) {
+                return Optional.empty();
+            }
+            if (!(args.get(0) instanceof SymbolReference)) {
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/sql/planner/DomainTranslator.java:956]: please extarct `args.get(0)` and `args.get(1)` as separate variables

File: presto-main/src/test/java/io/prestosql/sql/planner/TestDomainTranslator.java [hunk 2]
```diff
@@ -1582,6 +1582,30 @@ public void testLikePredicate()
         assertUnsupportedPredicate(not(like(C_VARCHAR, stringLiteral("abc\\_def"))));
     }
 
+    @Test
+    public void testStartsWithFunction()
+    {
+        Type varcharType = createUnboundedVarcharType();
+
+        // constant
+        testSimpleComparison(
```
Comment (sopel39) [inline presto-main/src/test/java/io/prestosql/sql/planner/TestDomainTranslator.java:1591]: add some edge cases like:
* empty pattern.
* pattern with trailing spaces
* pattern with escaped `.*` or `/`

File: presto-main/src/main/java/io/prestosql/sql/planner/DomainTranslator.java (hunk 1 above)
Comment (takezoe) [reply presto-main/src/main/java/io/prestosql/sql/planner/DomainTranslator.java]: fixed.

File: presto-main/src/test/java/io/prestosql/sql/planner/TestDomainTranslator.java (hunk 2 above)
Comment (takezoe) [reply presto-main/src/test/java/io/prestosql/sql/planner/TestDomainTranslator.java:1591]: Added empty pattern which is not supported in domain translation and `_` prefix pattern which is treated as a constant string as different from `LIKE`.

> - pattern with trailing spaces
> - pattern with escaped `.*` or `/`

What do these cases test? Unlike `LIKE` predicate, `STARTS_WITH` function always treats the second argument as  a constant string. It doesn't care whether characters are escaped or not.

Comment (sopel39) [reply presto-main/src/test/java/io/prestosql/sql/planner/TestDomainTranslator.java:1591]: > What do these cases test? Unlike LIKE predicate, STARTS_WITH function always treats the second argument as a constant string. It doesn't care whether characters are escaped or not.

You are right.

File: presto-main/src/test/java/io/prestosql/sql/planner/TestDomainTranslator.java [hunk 3]
```diff
@@ -1582,6 +1582,39 @@ public void testLikePredicate()
...
+                startsWith(C_VARCHAR, stringLiteral("_abc")),
+                C_VARCHAR,
+                startsWith(C_VARCHAR, stringLiteral("_abc")),
+                Domain.create(ValueSet.ofRanges(Range.range(varcharType, utf8Slice("_abc"), true, utf8Slice("_abd"), false)), false));
+
+        // empty
+        assertUnsupportedPredicate(startsWith(C_VARCHAR, stringLiteral("")));
+
+        // non-ASCII
+        testSimpleComparison(
```
Comment (sopel39) [inline presto-main/src/test/java/io/prestosql/sql/planner/TestDomainTranslator.java:1607]: Could you add test for `complement` case?

File: presto-main/src/test/java/io/prestosql/sql/planner/TestDomainTranslator.java [hunk 4]
```diff
@@ -1582,6 +1582,39 @@ public void testLikePredicate()
         assertUnsupportedPredicate(not(like(C_VARCHAR, stringLiteral("abc\\_def"))));
     }
 
+    @Test
+    public void testStartsWithFunction()
```
Comment (sopel39) [inline presto-main/src/test/java/io/prestosql/sql/planner/TestDomainTranslator.java:1586]: Could you add a test for generic function call (that it's not supported)?

File: presto-main/src/test/java/io/prestosql/sql/planner/TestDomainTranslator.java (hunk 3 above)
Comment (takezoe) [reply presto-main/src/test/java/io/prestosql/sql/planner/TestDomainTranslator.java:1609]: Added.

File: presto-main/src/test/java/io/prestosql/sql/planner/TestDomainTranslator.java (hunk 4 above)
Comment (takezoe) [reply presto-main/src/test/java/io/prestosql/sql/planner/TestDomainTranslator.java:1586]: Added `testUnsupportedFunctions()` which tests some unsupported functions.

**Comment by sopel39** [comment]
merged, thanks!
